pkg/kubefed2/federate/federate.go

```diff
@@ -0,0 +1,281 @@
+		Name:      j.typeName,
+	}
+	qualifiedResourceName := ctlutil.QualifiedName{
+		Namespace: j.resourceNameSpace,
+		Name:      j.resourceName,
+	}
+
+	return FederateResource(hostConfig, qualifiedTypeName, qualifiedResourceName)
+}
+
+func FederateResource(hostConfig *rest.Config, qualifiedTypeName, qualifiedResourceName ctlutil.QualifiedName) error {
+	typeConfig, err := lookUpTypeDetails(hostConfig, qualifiedTypeName)
+	if err != nil {
+		return err
+	}
+
+	templateResource, err := getTargetResource(hostConfig, typeConfig, qualifiedResourceName)
+	if err != nil {
+		return err
+	}
+
+	return createFedResource(hostConfig, typeConfig, templateResource)
+}
+
+func lookUpTypeDetails(config *rest.Config, qualifiedTypeName ctlutil.QualifiedName) (*fedv1a1.FederatedTypeConfig, error) {
+	fedClient, err := util.FedClientset(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get federation clientset")
+	}
+
+	typeConfig, err := fedClient.CoreV1alpha1().FederatedTypeConfigs(qualifiedTypeName.Namespace).Get(qualifiedTypeName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "FederatedTypeConfig %q, does not exist. Type needs to be enabled first", qualifiedTypeName)
+		}
+		return nil, errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", qualifiedTypeName)
+	}
+
+	crdClient, err := apiextv1b1client.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating crd client")
+	}
+
+	_, err = crdClient.CustomResourceDefinitions().Get(typeconfig.GroupQualifiedName(typeConfig.GetFederatedType()), metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving Federated<type> CRD corresponding to FederatedTypeConfig %q. Federated<type> CRD should exist", qualifiedTypeName)
+	}
+
+	return typeConfig, err
+}
+
+func getTargetResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTypeConfig, qualifiedResourceName ctlutil.QualifiedName) (*unstructured.Unstructured, error) {
+	targetAPIResource := typeConfig.GetTarget()
+	targetClient, err := ctlutil.NewResourceClient(hostConfig, &targetAPIResource)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating target resource client")
+	}
+
+	resource, err := targetClient.Resources(qualifiedResourceName.Namespace).Get(qualifiedResourceName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.New(fmt.Sprintf("Target resource %q not found", qualifiedResourceName))
+		}
+		return nil, errors.Wrap(err, "Error retreiving the target resource")
+	}
+
+	return resource, nil
+}
+
+func createFedResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTypeConfig, template *unstructured.Unstructured) error {
+	fedAPIResource := typeConfig.GetFederatedType()
+	fedClient, err := ctlutil.NewResourceClient(hostConfig, &fedAPIResource)
+	if err != nil {
+		return errors.Wrap(err, "Error creating federation resource client")
```

Comment:
-> `Error creating client for <federated type>`

Reply:
Updated!